File: kotlin/protocol/src/main/kotlin/org/hpb/protocol/Attestation.kt
package org.hpb.protocol

import org.hpb.engine.nostr.Events
import org.hpb.engine.nostr.NostrEvent
import org.hpb.protocol.Pj.l
import org.hpb.protocol.Pj.s

/**
 * Portable KYC/qualification attestations. KYC is OPTIONAL at the job level:
 * offers declare accepted attesters; workers without badges still see and
 * take open jobs. Attesters compete — no single gatekeeper.
 */
data class Attestation(
    val schema: String,
    val subject: String,
    val status: String, // "valid" | "revoked"
    val issuedAt: Long,
    val validUntil: Long,
)

object Attestations {
    fun toEvent(attesterPrivkey: ByteArray, a: Attestation, createdAt: Long): NostrEvent =
        Events.sign(
            attesterPrivkey, ProtocolKinds.ATTESTATION,
            tags = listOf(
                listOf("d", "${a.schema}:${a.subject}"),
                listOf("p", a.subject),
                listOf("t", a.schema),
            ),
            content = Pj.obj(
                "v" to Pj.num(ProtocolKinds.VERSION),
                "schema" to Pj.str(a.schema),
                "subject" to Pj.str(a.subject),
                "status" to Pj.str(a.status),
                "issued_at" to Pj.num(a.issuedAt),
                "valid_until" to Pj.num(a.validUntil),
            ).toString(),
            createdAt = createdAt,
        )

    fun fromEvent(event: NostrEvent): Attestation {
        require(event.kind == ProtocolKinds.ATTESTATION) { "not an attestation" }
        val content = Pj.parse(event.content)
        return Attestation(
            schema = content.s("schema"),
            subject = content.s("subject"),
            status = content.s("status"),
            issuedAt = content.l("issued_at"),
            validUntil = content.l("valid_until"),
        )
    }

    /**
     * Does this event satisfy an offer's KYC policy for the worker, now?
     * Verifies: event signature, accepted attester, subject binding,
     * valid status, and expiry. Fails closed on any gap.
     */
    fun satisfies(event: NostrEvent, policy: KycPolicy, worker: String, now: Long): Boolean {
        if (!policy.required) return true
        if (!isAcceptedAttester(event, policy)) return false
        val attestation = runCatching { fromEvent(event) }.getOrNull() ?: return false
        return attestation.subject == worker &&
            attestation.status == "valid" &&
            now < attestation.validUntil
    }

    private fun isAcceptedAttester(event: NostrEvent, policy: KycPolicy): Boolean =
        Events.verify(event) &&
            event.kind == ProtocolKinds.ATTESTATION &&
            event.pubkey in policy.attesters
}

File: ios/HpbCore/Sources/HpbCore/Attestation.swift
import Foundation

/// Portable KYC/qualification attestations. KYC is OPTIONAL at the job
/// level; attesters compete — no single gatekeeper.
public struct Attestation {
    public let schema: String
    public let subject: String
    public let status: String // "valid" | "revoked"
    public let issuedAt: Int64
    public let validUntil: Int64

    public init(schema: String, subject: String, status: String, issuedAt: Int64, validUntil: Int64) {
        self.schema = schema
        self.subject = subject
        self.status = status
        self.issuedAt = issuedAt
        self.validUntil = validUntil
    }
}

public enum Attestations {
    public static func toEvent(_ attesterPrivkey: [UInt8], _ a: Attestation, createdAt: Int64) throws -> NostrEvent {
        try Events.sign(
            attesterPrivkey, kind: ProtocolKinds.attestation,
            tags: [
                ["d", "\(a.schema):\(a.subject)"],
                ["p", a.subject],
                ["t", a.schema],
            ],
            content: Cj.write(Cj.obj([
                ("v", Cj.num(ProtocolKinds.version)),
                ("schema", .str(a.schema)),
                ("subject", .str(a.subject)),
                ("status", .str(a.status)),
                ("issued_at", Cj.num(a.issuedAt)),
                ("valid_until", Cj.num(a.validUntil)),
            ])),
            createdAt: createdAt
        )
    }

    public static func fromEvent(_ event: NostrEvent) throws -> Attestation {
        guard event.kind == ProtocolKinds.attestation else { throw HpbError.invalid("not an attestation") }
        let content = try Cj.parse(event.content)
        return Attestation(
            schema: try content.s("schema"),
            subject: try content.s("subject"),
            status: try content.s("status"),
            issuedAt: try content.l("issued_at"),
            validUntil: try content.l("valid_until")
        )
    }

    /// Does this event satisfy an offer's KYC policy for the worker, now?
    /// Fails closed on any gap.
    public static func satisfies(_ event: NostrEvent, policy: KycPolicy, worker: String, now: Int64) -> Bool {
        if !policy.required { return true }
        guard isAcceptedAttester(event, policy) else { return false }
        guard let attestation = try? fromEvent(event) else { return false }
        return attestation.subject == worker &&
            attestation.status == "valid" &&
            now < attestation.validUntil
    }

    private static func isAcceptedAttester(_ event: NostrEvent, _ policy: KycPolicy) -> Bool {
        Events.verify(event) &&
            event.kind == ProtocolKinds.attestation &&
            policy.attesters.contains(event.pubkey)
    }
}
